### DISCUSSION TITLE
Storing non-serializable data inside store

### DISCUSSION THREAD

**Discussion by dannobytes** [pmndrs/zustand#1873] [Q&A]
An important concept and limitation in Redux is that Redux stores can only hold serializable data such as primitives, arrays and objects. It does not know how to serialize objects like Date, regex, functions, Symbols, etc.

Does Zustand have similar limitations in terms of what types of data can or should be stored in it's state?

I couldn't find much documentation speaking to this. But it seems that while Zustand does not explicitly disallow non-serializable objects as state values, I wonder if it would still be best practice not to? Or perhaps only if you use the "persist" middleware, you just need to make sure to write custom serializing functions to handle them properly?

What about when using redux devtools? How would that extension handle non-serializable values? Would it break or simply be omitted?

Curious to see some recommendation that speaks to this on the website docs too!

**Comment by dai-shi** [comment]
> Curious to see some recommendation that speaks to this on the website docs too!

Would be a good challenge for the docs team.

Fun fact: `increasePoplulation` is non serializable.
https://github.com/pmndrs/zustand/blob/a53433543cf36f6379242f3212363eb770340dd1/readme.md?plain=1#L32

**Reply by dannobytes** [reply]
Didn't know there was a separate docs team 😄 

And very true about the `function` types. I did notice that, which made me question my own assumption that non-serializable data was not allowed. In fact, it doesn't seem like Zustand cares all that much. So you could in theory create React ref properties in the store too.

Curious what other kinds of non-serializable values other devs have stored...

**Reply by dai-shi** [reply]
> it doesn't seem like Zustand cares all that much.

Correct.

> So you could in theory create React ref properties

Yeah.

> Didn't know there was a separate docs team

We discuss in this discussion: #1033

**Comment by dannobytes** [comment]
I wanted to follow up with some additional findings since testing this out with non-standard, non-serializable JS objects such as `Date`, `HTMLElement` and `React.Ref`s.

### `Date` objects are 👌🏼

`Date` objects are serializable by default into a string, e.g. `Date().toString()`. So these are totally acceptable types with no complications and work well with both `immer` and `devtools` middlewares.

### `HTMLElement`s is problematic for both `immer` and `devtools` middlewares in two distinct ways.
1. With `immer`, the types for `WritableDraft<HTMLDivElement>` and `HTMLDivElement` seem to be entirely incompatible
<img width="1504" alt="image" src="https://github.com/pmndrs/zustand/assets/751279/83410929-fb34-44e3-a702-232d8351d297">


2. With `devtools`, Redux DevTools extension cannot handle serializing HTML element objects as state or action payloads. It crashes as a result.
<img width="652" alt="image" src="https://github.com/pmndrs/zustand/assets/751279/c0132fc2-a600-442d-9290-a81ffa1d24d9">


### `React.Ref` objects as state properties are problematic

Naturally, ref objects are designed to be mutable with its internal `current` property needing to be updated while keeping the parent object the same. When attempting to do this with a store state property that is a React.Ref, you get an immutability error b/c an attempt to mutate the ref object took place.

e.g. 
```ts
// store/index.ts
export const store = createStore()(immer(devtools((set, get) => {
  return {
    ref: React.createRef(),
    ...
  }
})));
```

```ts
// Component.tsx
function Component() {
  const storeRef = useStore(s => s.ref);
  return <div ref={storeRef}>...</div>
}
```

<img width="667" alt="image" src="https://github.com/pmndrs/zustand/assets/751279/5a4d413e-40bc-46ee-992c-2c9bb9effb15">


### Conclusion

My conclusion thus far is that while Zustand does not inherently prevent or provide any proactive guard rails against using complex, non-serializable objects like HTML elements, React refs and/or others, there are genuine road blocks ahead from adjacent tooling such as immer and devtools middleware.

I would definitely recommend adding some documentation about this, mildly steering devs away from it.

**Reply by dai-shi** [reply]
For devtools, you might want to check out `serialize` option. https://github.com/reduxjs/redux-devtools/blob/main/extension/docs/API/Arguments.md#serialize

**Reply by dai-shi** [reply]
For createRef and immer, it seems to work in a simple form: https://codesandbox.io/s/nameless-water-vwty7r?file=/src/App.js
